adrs/0049-prebuilt-rubies-at-runtime.md

```diff
@@ -0,0 +1,75 @@
+
+This is in contrast to other setup actions like setup-node where the cache is an optimization but supports pulling any version by a specified semver at runtime.
+
+## Goals
+
+- Offer as many versions as possible by platforms
+- Offer popular Ruby variants like JRuby
+- Consistent with other setup-xxx actions patterns and Actions workflows.
+
+## Decisions
+
+### Prebuilt Rubies
+
+Other tools like offer a [distribution](https://nodejs.org/dist/) which offers a [queryable endpoint](https://nodejs.org/dist/index.json) which allows the desired version [by a version spec which is semver](https://github.com/actions/setup-node#setup-node).
+
+Other CI services like Travis support pulling a wide variety of [ruby versions offered here](http://rubies.travis-ci.org/).
+
+Requirements:
+- Offer prebuilt Ruby, JRuby, TruffleRuby versions 
+- Setup-Ruby exposes `ruby-version` which accepts a semver
+- Rubies are discoverable and queryable by the Setup-Ruby action.
+- Initially offer for all platforms offered by the hosted Actions VMs (Ubuntu 18, Mac 10.15 and Windows 2019)
+
+NOTE: The prebuilt rubies offered will only be supported for use by GitHub Actions.
+
+The scripts and tooling to build the Rubies will be at `actions/build-ruby` and it will be consumed by the existing action `actions/setup-ruby`.  Since the capabilities are additive, there's no need to create a `v2`.  There is no compat break.
+
+A specific version of a Ruby will be offered as an individual semi-immutable build-ruby repo release similar to how the [actions runner exposes versions](https://github.com/actions/runner/releases/tag/v2.164.0) where each version offers n platforms.  This is also similar to [Travis individual versions](http://rubies.travis-ci.org/ubuntu/18.04/s390x/ruby-2.6.5).
+
+NOTE: semi-immutable means it's desirable to be immutable but it's possible to patch.  Note that 
+
+Option 1: GitHub GPR Universal Package
+
+The `actions/build-ruby` repo offers packages for [example](https://github.com/actions/setup-ruby/packages). 
+
+Option 2: GitHub Release
+
+The `actions/build-ruby` repo offers releases for [example](https://github.com/actions/runner/releases/tag/v2.164.0)
+
+NOTE: release assets are backed by a CDN
+
+Offering each version as an individual package / release offers queryable APIs and the ability to convey whether pre-release or not by version using the packaging and release features.
+
+This is fairly straight forward to come up with a scheme to complete automate with a workflow.
+
+### Setup action
+
+Repo releases are also queryable via an [http api](https://developer.github.com/v3/repos/releases/).  This allows the `actions/setup-ruby` action to query the versions. match the semver version spec against the list and resolve the latest matching the spec.
```

Comment:
Getting the list of releases via https://developer.github.com/v3/repos/releases/#list-releases-for-a-repository might take some time, especially if there are many releases, the JSON responses are paginated, and therefore multiple requests are issued.

Reply:
Yes, another reason to use packages.  However, in practice I don't believe it will be a problem.  Note that the tool-cache lib queries the cache first and the VM would populate latest versions so it short circuits and never queries even if the cache can satisfy the semver.  Also, exact version specs don't need to query at all.

Reply:
I forwarded this to our teams that gen images and also our packaging folks.  There's other work going on in these areas so it would be great to drive to consensus with community ❤️